Add test cases for recent ES2016 fixes

anba opened the pull request:
- lastIndex in RegExp.prototype[Symbol.split]: tc39/ecma262@08b4756747f9c967058239df627da544412740a6
- Missing number conversion in DataView.prototype.setXXX: tc39/ecma262@4f875fe96dcbd5e6363dd3d5b510cd5109199f5b
- Negative zero byteoffset in TypedArray: tc39/ecma262@2d1ed20db778e8f69f66ba76b351b9322fd38495
- EvalDeclarationInstantiation throws TypeError: tc39/ecma262@2be6968715946a4763f6ca8633e311ab7ce63577
- BindingRestElement allows BindingPattern: tc39/ecma262@d322357e6be95bc4bd3e03f5944a736aac55fa50
- Eval in parameters with computed property keys: tc39/ecma262@04e2e9b7197a33612202e85065e3f8d8385fbcef
- Use strict with non-simple parameters: tc39/ecma262@15b0db41edd0e519e94668bf13765fe458840766
- \__proto__ in strict mode: tc39/ecma262@5c1984334d6d42538527bd022e68bfbfee5cb652

goyakin commented on test/built-ins/DataView/prototype/setFloat32/index-to-integer.js:
```diff
@@ -0,0 +1,48 @@
+var dataView = new DataView(new ArrayBuffer(8));
+
+dataView.setFloat32(+0, 1);
+assert.sameValue(dataView.getFloat32(0), 1, "setFloat32(null, 1)");
```
**null** should be **+0** instead. The index should be updated in the next two assertion messages too.

goyakin commented on test/built-ins/DataView/prototype/setFloat64/index-to-integer.js:
```diff
@@ -0,0 +1,48 @@
+var dataView = new DataView(new ArrayBuffer(16));
+
+dataView.setFloat64(+0, 1);
+assert.sameValue(dataView.getFloat64(0), 1, "setFloat64(null, 1)");
```
Same as `setFloat32`, the assertion messages should be updated.

goyakin commented on test/built-ins/DataView/prototype/setFloat32/index-check-before-value-conversion.js:
```diff
@@ -0,0 +1,73 @@
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+es7id: 24.2.1.2
```
I think the frontmatter should mention `DataView.prototype.setFloat32` instead of `SetViewValue`.

goyakin commented on test/built-ins/DataView/prototype/setInt16/index-to-integer.js:
```diff
@@ -0,0 +1,48 @@
+var dataView = new DataView(new ArrayBuffer(8));
+
+dataView.setInt16(+0, 1);
+assert.sameValue(dataView.getInt16(0), 1, "setInt16(null, 1)");
```
Same as `setFloat32`, the assertion messages should be updated.

goyakin commented on test/built-ins/DataView/prototype/setInt32/index-to-integer.js:
```diff
@@ -0,0 +1,48 @@
+var dataView = new DataView(new ArrayBuffer(8));
+
+dataView.setInt32(+0, 1);
+assert.sameValue(dataView.getInt32(0), 1, "setInt32(null, 1)");
```
Same as `setFloat32`, the assertion messages should be updated.

goyakin commented on test/built-ins/DataView/prototype/setFloat32/index-check-before-value-conversion.js:
```diff
@@ -0,0 +1,73 @@
+  ...
+  3. Let numberIndex be ToNumber(requestIndex).
+  4. Let getIndex be ? ToInteger(numberIndex).
+  5. If numberIndex ≠ getIndex or getIndex < 0, throw a RangeError exception.
```
@anba, what do you think about adding a test case with an object coerced to an invalid number?

goyakin commented on test/built-ins/DataView/prototype/setFloat32/index-to-integer.js:
```diff
@@ -0,0 +1,48 @@
+
+  ...
+  3. Let numberIndex be ToNumber(requestIndex).
+  4. Let getIndex be ? ToInteger(numberIndex).
```
Adding a test case with an object coerced to a valid integer would make it more complete.

goyakin commented:
Thanks @anba. I left some minor comments inline, but it looks good to me otherwise.

anba commented:
Rebased and addressed review comments:
- Fixed assertion messages
- Added tests for DataView.prototype.setXXX(validNumberAsObject, value)
- Added tests for DataView.prototype.setXXX(invalidNumberAsObject, value)
- Updated frontmatter id to mention DataView.prototype methods instead of SetViewValue

Thanks for reviewing @goyakin!
